### PR TITLE
gh-127847: Fix position in the special-cased zipfile seek

### PR DESCRIPTION
**Pull request by dimaryaz**
* Issue: gh-127847

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2024-12-12-07-27-51.gh-issue-127847.ksfNKM.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix the position when doing interleaved seeks and reads in uncompressed unencrypted zip files
```
Comment (ZeroIntensity) [inline Misc/NEWS.d/next/Library/2024-12-12-07-27-51.gh-issue-127847.ksfNKM.rst]: It's good to have some context for what was fixed (also some typos; english is hard).
```suggestion
Fix the position when doing interleaved seeks and reads in uncompressed, unencrypted zip files returned by :meth:`zipfile.ZipFile.open`.
```

File: Lib/test/test_zipfile/test_core.py [hunk 2]
```diff
@@ -2333,6 +2333,18 @@ def test_read_after_seek(self):
                 fp.seek(1, os.SEEK_CUR)
                 self.assertEqual(fp.read(-1), b'men!')
 
+    def test_uncompressed_interleaved_seek_read(self):
+        # Issue 127847: Make sure the position in the archive is correct
```
Comment (ZeroIntensity) [inline Lib/test/test_zipfile/test_core.py]:
```suggestion
        # gh-127847: Make sure the position in the archive is correct
```

File: Lib/test/test_zipfile/test_core.py [hunk 3]
```diff
@@ -2333,6 +2333,18 @@ def test_read_after_seek(self):
...
+        # in the special case of seeking in a ZIP_STORED entry.
+        with zipfile.ZipFile(TESTFN, "w") as zipf:
+            zipf.writestr("a.txt", "123")
+            zipf.writestr("b.txt", "456")
+        with zipfile.ZipFile(TESTFN, "r") as zipf:
+            with zipf.open("a.txt", "r") as a, zipf.open("b.txt", "r") as b:
+                a.read(1)
+                b.seek(1)
```
Comment (ZeroIntensity) [inline Lib/test/test_zipfile/test_core.py]: Both of these can return values, should we assert that they're something?

**Review by ZeroIntensity — COMMENTED** [review]
This is a great first contribution, thanks :)

I have a few nitpicks, mostly related to some high-level triage tidbits.

File: Lib/test/test_zipfile/test_core.py (hunk 3 above)
Comment (dimaryaz) [reply Lib/test/test_zipfile/test_core.py]: Sure, fixed

**Comment by danifus** [comment]
Nice tests and the bug report was great.

The fix I suggested was incomplete (I forgot `_fileobj` was an instance of `_SharedFile` until after I responded to you). The fix needs to go up a level of abstraction :p

`_SharedFile` should manage the position within the zip entry for us but it's `seek` implementation isn't resuming from its current position (like it does in `read`):

https://github.com/python/cpython/blob/8bbd379ee30db0320ec3d31c37aee2a503902b0f/Lib/zipfile/__init__.py#L816-L835

`seek` should include something like `self._file.seek(self._pos)` like `read` does. This was hidden before the special cased `ZIP_STORED` seek as `ZipExtFile.seek` would use `read` to advance to the position if required (or seek back to the start of the file and read from there).

**Comment by dimaryaz** [comment]
Ah, got it.

I used an if/else, just to avoid doing two seeks unnecessarily - do you think that's better?

(Though an extra seek worked fine, too, so I can change it to that.)

**Comment by danifus** [comment]
I like the single seek you've done. Nice work!

(The CI tool failed for macos-13 but that looks unrelated to these changes. The failing test was: `test.test_multiprocessing_forkserver.test_processes`. I'm not sure if it needs to be rerun before it can be merged?)

**Review by danifus — APPROVED** [review]

**Comment by dimaryaz** [comment]
Just rebased, and everything passed now.

**Review by ZeroIntensity — APPROVED** [review]
LGTM, thanks for the fix :)

**Review by jaraco — APPROVED** [review]
LGTM. Thanks for the clean contribution and others for the thorough review.

**Comment by ZeroIntensity** [comment]
Congrats on your first contribution @dimaryaz

**Comment by dimaryaz** [comment]
Thanks everyone!

**Pull request merged**